homeassistant/components/knx/config_flow.py

```diff
@@ -0,0 +1,265 @@
+from xknx.io.gateway_scanner import GatewayDescriptor, GatewayScanner
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import (
+    CONF_KNX_AUTOMATIC,
+    CONF_KNX_CONNECTION_TYPE,
+    CONF_KNX_INDIVIDUAL_ADDRESS,
+    CONF_KNX_ROUTING,
+    CONF_KNX_TUNNELING,
+    DOMAIN,
+)
+from .schema import ConnectionSchema
+
+CONF_KNX_GATEWAY: Final = "gateway"
+
+
+class FlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a KNX config flow."""
+
+    VERSION = 1
+
+    _tunnels: list
+
+    async def async_step_user(self, user_input: dict | None = None) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        self._tunnels = []
+        return await self.async_step_type()
+
+    async def async_step_type(self, user_input: dict | None = None) -> FlowResult:
+        """Handle connection type configuration."""
+        errors: dict = {}
+        supported_connection_types = [CONF_KNX_TUNNELING, CONF_KNX_ROUTING]
+        fields = {}
+
+        if user_input is None:
+            gateways = await scan_for_gateways()
+
+            if gateways:
+                supported_connection_types.insert(0, CONF_KNX_AUTOMATIC)
+                self._tunnels = [
+                    gateway for gateway in gateways if gateway.supports_tunnelling
+                ]
+
+            fields = {
+                vol.Required(CONF_KNX_CONNECTION_TYPE): vol.In(
+                    supported_connection_types
+                )
+            }
+
+        if user_input is not None:
+            connection_type = user_input[CONF_KNX_CONNECTION_TYPE]
+            if connection_type == CONF_KNX_AUTOMATIC:
+                return self.async_create_entry(
+                    title=CONF_KNX_AUTOMATIC, data=user_input
+                )
+
+            if connection_type == CONF_KNX_ROUTING:
+                return await self.async_step_routing()
+
+            if connection_type == CONF_KNX_TUNNELING and self._tunnels:
+                return await self.async_step_tunnel()
+
+            return await self.async_step_manual_tunnel(user_input)
+
+        return self.async_show_form(
+            step_id="type", data_schema=vol.Schema(fields), errors=errors
+        )
+
+    async def async_step_manual_tunnel(
+        self, user_input: dict | None = None
+    ) -> FlowResult:
+        """General setup."""
+        errors: dict = {}
+
+        # prefill data for frontend
+        _gateway_ip: str = ""
+        _gateway_port: int = 3675
+        if user_input is not None and CONF_HOST in user_input:
+            _gateway_ip = user_input[CONF_HOST]
+            _gateway_port = user_input[CONF_PORT]
+
+        if user_input is not None and CONF_KNX_INDIVIDUAL_ADDRESS in user_input:
```

Comment:
`CONF_KNX_INDIVIDUAL_ADDRESS` is required, wouldn't it always be in `user_input` ?

Reply:
It's called in `async_step_tunnel` which allows the user to select a gateway from a list of gateways.

```
return await self.async_step_manual_tunnel(
                {
                    CONF_HOST: gateway.ip_addr,
                    CONF_PORT: gateway.port,
                    CONF_KNX_CONNECTION_TYPE: CONF_KNX_TUNNELING,
                }
            )
```

There the individual address will not be set, thus this check.

Reply:
I adjusted it to use instance variables as suggested